substrate/frame/bounties/src/lib.rs

```diff
@@ -174,6 +175,11 @@ pub enum BountyStatus<AccountId, BlockNumber> {
 		/// When the bounty can be claimed.
 		unlock_at: BlockNumber,
 	},
+	/// The bounty is approved with curator assigned.
+	ApprovedWithCurator {
```

Comment:
@muharem Because there still needs some time to pass when bounty is funded because as I understand only at beginning of the block bounties are funded. There could be a time when bounty is approved but funds are not available for some time. So, if we were to jump to a Funded status we would be trying to fund the bounty right away taking from the treasury, and if we don't have funds we can't fund bounty and propose a curator yet. So we have status that bounty is ApprovedWithCurator to be able to approve the bounty without having funds yet.

Correct me if I'm wrong in this

Reply:
you are right

Reply:
I’m not very familiar with the code, but having multiple statuses with similar meanings seems like a potential code smell.

How about modifying `Approved` with the field `maybe_curator: Option<AccountId>`? Probably we want to avoid any breaking change (which is totally understandable)? 

I leave the judgement to you guys. Though it would be great to have some doc explaining expected lifecycle of a bounty (I couldn't find one).

Reply:
Hello, this is Vasilis.
This will not be a huge deal for us on the Bounty Manager, we will just add a new status once the change is implemented, or we could also modify the current one.
As we are not running an indexer, it should in theory be a small effort for us to update it.

Reply:
the question here - should we alter existing `Approve` status by adding an optional curator or we add new `ApproveWithCurator` status. we trying to learn here from you, from clients of this API. as a code maintainer we rather alter `Approve` to keep things cleaner (but I believe the impact is low). however I think from the client perspective it is better we add new status (append new enum variation). in this case it should not be breaking change (I believe the impact is much higher).

Reply:
In my opinion, it is cleaner to alter it as well and for us it's not that much of a problem, as it will not be really a big breaking change for us because our app is pretty small and we are not indexing.

What I can imagine is that it will be a more important and/or breaking change for other platforms who might be indexing or saving this status, like e.g. Polkassembly and Subsquare. It might be actually a good idea to ask their opinions as well.

Reply:
Indexers is good example. Can you link someone from those teams?
But even for your use case, is this a breaking change? I can also imagine it to be a non-breaking, if metadata automatically updated and types decoded from it.

Reply:
I will ping them, as I don't know their handles.
I think even if the metadata updates and gets decoded, it will still be a "soft" break, as we will need to update some thigns to show it correctly. Most likely unless a bounty is currently in that status before we update it, I believe it won't have a breaking effect though.

Reply:
Adding new logic is better than alteration to data structure, because it cost less work to add patches.

Reply:
Thank you.
It would be helpful to create a document (or perhaps even a talk, followed by a document) to give runtime developers a clearer understanding of existing client use cases, what constitutes a breaking change for a client (less and more painful), and which changes are worthwhile. It could clarify what is generated dynamically from metadata and what isn’t, as well as whether enum variant indexes or names are used. This will likely require collaborative efforts to cover more use cases.
cc: @bill-pap @wliyongfeng @Tbaut @josepot

Reply:
> But my thought was - better non breaking change, all clients keep working. new feature available only to the clients adopting the new status.

Generally speaking, breaking changes should be avoided if possible (when discussing downstream clients, and tools), and if they are necessary then they should be correctly communicated in a changelog - and/or runtime release docs.

> It would be helpful to create a document (or perhaps even a talk, followed by a document) to give runtime developers a clearer understanding of existing client use cases, what constitutes a breaking change for a client (less and more painful)

Even though this would be a tall task, I totally agree. I would love to see more documentation on this front.

> It could clarify what is generated dynamically from metadata and what isn’t

Atleast from the perspective of PJS - everything is dynamically generated from the metadata now. That being said the static typing  (not to be confused with the dynamic generation at runtime) requires a few longer steps to get the types reflected in the PJS libraries - generate the static types - fix any required changes - release the api - then update all deps downstream. 

If I am not mistaken the PAPI descriptors do a way better job at supplying the types (But I am no expert - yet :))

Reply:
If we want to make this effort future-proof, and bring it up to understanding with #4722, we should be asking ourselves: 

`BountyStatus` is the status enum that the internal logic of the pallet would need to care about. Is it even reasonable to expose this to the users? 

For example, if we are to design a `fn status_of(bounty) -> BountyStatus`, what would the return type be?

Reply:
> How about modifying `Approved` with the field `maybe_curator: Option<AccountId>`?

I'm sorry that I'm late to the party. Modifying `Approved` to be an `Option<AccountId>` would have been better IMO.

The reason is simple: if someone is using code that doesn't take into account the new state, then if they receive the variant "`ApprovedWithCurator`" their DApp will blow up no matter what, because they don't know how to deal with that variant.

On the other hand, modifying the `Approved` variant to support an optional value, may prevent some DApps from blowing up. Also, at least with PAPI, it makes it much simpler for DApp developers to adapt to the change on their DApps.

I'm sorry that I'm sharing this after the PR has been merged.

Reply:
This means the impact can be different for the dapps and indexers. We need to learn more about it and document. It seems like we really can improve the status quo.

Reply:
I really fail to understand why this change is better for indexers.